proposals/0000-backpack.rst

```diff
@@ -0,0 +1,1030 @@
+by other indefinite libraries which depend on it.
+
+An indefinite library can be used in two ways:
+
+1. It can be depended upon by other libraries, which inherit
+   the requirement of this library::
+
+        -- in the Cabal file
+        library stringutils-indef
+            build-depends: concat-indef
+            exposed-modules: StringUtils
+
+        -- in StringUtils.hs
+        module StringUtils where
+        import Concat
+        import Str -- the signature is importable too
+
+   An inherited requirement can be extended simply by adding
+   a local signature.  Similarly, if two dependencies have
+   the same requirement, the two requirements are merged
+   together (*interface composibility.*)
+
+   Using these facilities, a user can, for example, write a **signature
+   library** simply by defining a library which has only signatures in
+   it, which allow signatures to be shared between multiple indefinite
+   libraries.
+
+2. It can be **instantiated** (possibly multiple times)
+   into a **definite library** which can be compiled. (*Extensibility*)
+   A client instantiates a library by bringing it into scope with
+   another library which exposes modules with the same names as the
+   signatures (a process called **mix-in linking**)::
+
+        library str-bytestring
+            exposed-modules: Str
+
+        library concat-bytestring
+            build-depends: str-bytestring, concat-indef
+            reexported-modules:
+                -- Concat from concat-indef is instantiated
+                -- with Str from str-bytestring.  We can
+                -- reexport it under a qualified name for
+                -- more convenient use.
+                Concat as Concat.ByteString
+
+   Compilation does not occur until the implementation of all
+   signatures is known, so that when compilation occurs, a compiler can optimize
+   as if Backpack was not present (*no performance overhead.*)
+   The types for each instantiation are considered distinct
+   (in a sense that is made precise in the `Identifiers`_ section.)
+
+Backpack additionally provides the ``backpack-includes`` field for
+renaming modules to make their names coincide (so that instantiation
+can take place.)
+
+Identifiers
+-----------
+
+In this section, we describe the grammar of identifiers in Backpack.
+These identifiers are used, for example, to determine the unique
+name associated with each entity in Haskell, which in turn determines
+when two types are equal.
+
+::
+
+    ComponentId ::= [A-Za-z0-9-_.]+
+    UnitId      ::= ComponentId
+                  | ComponentId "[" ModuleSubst "]"
+    ModuleSubst ::= ( ModuleName "=" Module ) +
+    -- from Haskell'98
+    ModuleName  ::= [A-Z][A-Za-z0-9_']* ( "." [A-Z][A-Za-z0-9_']* ) +
+    Module      ::= UnitId ":" ModuleName
+                  | "<" ModuleName ">"
+
+.. _ComponentId:
+
+A **component identifier** consists of an arbitrary sequence of
```

Comment:
Intuitively, the transitive closure of the source code. Where we mean including the direct dependencies, but not the indirect dependencies.